# Relatório 2

**Nome:** Thiago Lopes <br>
**Matrícula:** 20100358 <br>
**Turma:** T2

# Bibliotecas python utilizadas

In [73]:
import numpy as np
import math

# Funçoes Auxiliares

In [74]:
def F_cos_theta(x, F, theta):
    return F * math.cos(theta)

def exact_solution_B(x):
    return -1 + 2 * np.exp(x) - x

# Métodos implementados: 

In [75]:
def simpson_13(a, b, n, f, F_values, theta_values):
    if n == 1:
        h = b - a
        simp13 = 2 * h * (f(a, F_values[0], theta_values[0]) + 4 * f((a + b) / 2, F_values[1], theta_values[1]) + f(b, F_values[2], theta_values[2])) / 6
    else:
        h = (b - a) / n
        
        sum = f(a, F_values[0], theta_values[0]) + f(b, F_values[-1], theta_values[-1])
        
        for i in range(1, n):
            x = a + i * h
            
            if i % 2 == 0:
                sum += 2 * f(x, F_values[i], theta_values[i])
            else:
                sum += 4 * f(x, F_values[i], theta_values[i])
                
        simp13 = h * sum / 3
    
    return simp13

In [76]:
def trapezio(a, b, n, f, F_values, theta_values):
    if n == 1:
        h = b - a
        trap = h * (f(a, F_values[0], theta_values[0]) + f(b, F_values[1], theta_values[1])) / 2
    else:
        h = (b - a) / n
        
        sum = f(a, F_values[0], theta_values[0])
        
        for i in range(1, n):
            x = a + i * h
            sum += 2 * f(x, F_values[i], theta_values[i])
            
        sum += f(b, F_values[-1], theta_values[-1])
        
        trap = h * sum / 2
    
    return trap
    

In [77]:
def simpson_38(a, b, n, f, F_values, theta_values):
    if n == 1:
        h = b - a
        simp38 = 3 * h * (f(a, F_values[0], theta_values[0]) + 3 * f((2 * a + b) / 3, F_values[1], theta_values[1]) + 3 * f((a + 2 * b) / 3, F_values[2], theta_values[2]) + f(b, F_values[3], theta_values[3])) / 8
    else:
        h = (b - a) / n
        
        sum = f(a, F_values[0], theta_values[0]) + f(b, F_values[-1], theta_values[-1])
        
        for i in range(1, n):
            x = a + i * h
            
            if i % 3 == 0:
                sum += 2 * f(x, F_values[i], theta_values[i])
            else:
                sum += 3 * f(x, F_values[i], theta_values[i])
                
        simp38 = 3 * h * sum / 8

    return simp38

In [78]:
def euler_method(f, a, b, y0, m):
    h = (b - a) / m
    x = a
    y = y0
    
    VetX = [x]
    VetY = [y]
    
    for i in range(1, m + 1):
        fxy = f(x, y)
        x = a + i * h
        y = y + h * fxy
        
        VetX.append(x)
        VetY.append(y)
    
    return VetX, VetY

# Exercícios

## Exercício 1

1) O trabalho realizado por uma força F(x) cujo ângulo entre a direção do movimento e a força é dado por θ(x), pode ser obtido pela seguinte fórmula: 

W =∫^xn^x0 F( x ) cos(θ ( x )) dx

Em que x0 e xn são as posições inicial e final, respectivamente.

Calcule a melhor aproximação ao trabalho realizado, W, ao puxar um bloco da posição 0 até à posição 30 sabendo que a força aplicada e o ângulo usado são dados na tabela seguinte.

x    | 0   | 2,5 | 5   | 15  | 20   | 25   | 30  |<br>
F(x) | 0   | 7   | 9   | 14  | 10,5 | 12   | 5   |<br>
θ(x) | 0,5 | 0,9 | 1,4 | 0,9 | 1,3  | 1,48 | 1,5 |<br>

Observação: Verificam-se 3 grupos de espaçamentos distintos. <br>

a) Aplicar o Método de 1/3 de Simpson Repetido no Grupo 1, o qual possui 2 subintervalos, h = 2,5.<br>

b) Aplicar o Método do Trapézio Repetido no Grupo 2, o qual possui 1 subintervalo, h = 10.<br>

c) Aplicar o Método do 3/8 Simpson Repetido no Grupo 3, o qual possui 3 subintervalos, h = 5.<br>


In [79]:
# Valores fornecidos
x_values = [0, 2.5, 5, 15, 20, 25, 30]
F_values = [0, 7, 9, 14, 10.5, 12, 5]
theta_values = [0.5, 0.9, 1.4, 0.9, 1.3, 1.48, 1.5]
            
# Metodo de Simpson no intervalo [0, 2.5, 5] com 2 subintervalos
a, b = 0, 5
n = 2
resultado1 = simpson_13(a, b, n, F_cos_theta, F_values[:3], theta_values[:3])
print('\nSimpson 1/3')
print(f'Trabalho aproximado no intervalo [0, 2.5, 5]: {resultado1}')

# ------------------

# Metodo do Trapézio no intervalo [5, 15] com 1 subintervalos
a, b = 5, 15
n = 1
resultado2 = trapezio(a, b, n, F_cos_theta, F_values[2:4], theta_values[2:4])
print('\nTrapezio')
print(f'Trabalho aproximado no intervalo [5, 15]: {resultado2}')

# -------------------

# Metodo de Simpson 3/8 no intervalo [15, 20, 25, 30] com 3 subintervalos
a, b = 15, 30
n = 3  # Três subintervalos
resultado3 = simpson_38(a, b, n, F_cos_theta, F_values[3:], theta_values[3:])
print('\nSimpson 3/8')
print(f'Trabalho aproximado no intervalo [15, 20, 25, 30]: {resultado3}\n')

sum123 = resultado1 + resultado2 + resultado3

print(f"Soma dos Trabalhos Calculados: {sum123}")



Simpson 1/3
Trabalho aproximado no intervalo [0, 2.5, 5]: 15.778986164733977

Trapezio
Trabalho aproximado no intervalo [5, 15]: 51.16121920945736

Simpson 3/8
Trabalho aproximado no intervalo [15, 20, 25, 30]: 38.899907149720256

Soma dos Trabalhos Calculados: 105.84011252391159


Método de 1/3 de Simpson Repetido: Para o intervalo [0, 2,5, 5], o trabalho aproximado foi de 15,78. Este método foi eficiente para intervalos com poucos pontos, proporcionando uma boa aproximação devido a forma parabólica da função.<br>

Para o intervalo [5, 15], que possui apenas 1 subintervalo com h = 10, foi aplicado o Método do Trapézio. Esse método é mais simples e adequado para casos onde a função não é muito irregular, mas pode não é tão preciso quanto os métodos de Simpson para intervalos com maior variação.<br>

No intervalo [15, 20, 25, 30], aplicamos o Método do 3/8 de Simpson Repetido, que é uma variante do Método de Simpson e é particularmente útil quando há mais pontos de dados e a função apresenta curvas mais complexas.<br>

Método de Simpson, tanto na forma de 1/3 quanto de 3/8, tende a ser mais preciso em intervalos com mais subintervalos e onde a função apresenta maior curvatura, enquanto o Método do Trapézio é mais simples e direto, adequado para intervalos com menos variação.

A soma dos trabalhos calculados reflete de forma geral o trabalho total realizado, indicando que os métodos utilizados foram eficazes na aproximação do trabalho realizado.<br>

## Exercício 2

2) Para os problemas de valor inicial<br>

A)  y’ = x^2 + y^2<br>
    y(0) = 0<br>
    
B)  y’ = x + y<br>
    y(0) = 1<br>

a) Calcule y(0,5) para o problema A e y(0,2) para o problema B usando o Método de Euler. Considere h = 0,1 e h = 0.05.<br>

b) Compare os resultados obtidos na letra a com a solução exata do problema A e B<br>

In [80]:
f_a = lambda x, y: x**2 + y**2
f_b = lambda x, y: x + y

# parâmetros iniciais
x0_A, y0_A = 0, 0
x0_B, y0_B = 0, 1
h_A = 0.1
h_B = 0.05

# Número de passos (m)
m_A = int(0.5 / h_A)
m_B = int(0.2 / h_B)

VetX_A, VetY_A = euler_method(f_a, x0_A, 0.5, y0_A, m_A)
VetX_B, VetY_B = euler_method(f_b, x0_B, 0.2, y0_B, m_B)

# Valores aproximados
y_0_5_A = VetY_A[-1]
y_0_2_B = VetY_B[-1]

# Solucao Exata
y_exact_B = exact_solution_B(0.2)

print(f"y(0.5) para o problema A usando o Método de Euler: {y_0_5_A}")
print(f"y(0.2) para o problema B usando o Método de Euler: {y_0_2_B}")
print(f"\nSolução exata de y(0.2) para o problema B: {y_exact_B}")

# Erro relativo para o problema B
erro_relativo_B = abs((y_0_2_B - y_exact_B) / y_exact_B) * 100
print(f"\nErro relativo para o problema B: {erro_relativo_B:.2f}%")

y(0.5) para o problema A usando o Método de Euler: 0.030022207380957063
y(0.2) para o problema B usando o Método de Euler: 1.2310124999999998

Solução exata de y(0.2) para o problema B: 1.2428055163203398

Erro relativo para o problema B: 0.95%


No problema A, calculamos y(0,5) usando o Método de Euler e obtivemos o valor de 0.030022207380957063. Como não possuímos a solução exata para comparação, não podemos avaliar diretamente a precisão do método para esse caso. No entanto, sabemos que o Método de Euler pode introduzir erros significativos, especialmente para equações diferenciais não lineares, como é o caso de y’ = x^2 + y^2.<br>

No problema B, comparamos o resultado obtido pelo Método de Euler com a solução exata. Utilizando o Método de Euler, encontramos y(0,2)=1.2310124999999998. A solução exata, y(0,2)=1.2428055163203398, apresenta um valor um pouco maior.<br>

O erro relativo é de 0.95%, o que indica que o Método de Euler fornece uma boa aproximação para valores pequenos de x, apesar de ainda apresentar uma pequena discrepância. Isso é esperado, pois o Método de Euler é um método numérico de primeira ordem, e a precisão pode ser aumentada utilizando métodos de maior ordem ou reduzindo o tamanho do passo h.<br>

Portanto, os resultados demonstram que o Método de Euler é uma ferramenta útil para encontrar soluções aproximadas de equações diferenciais, embora deva-se ter cuidado com o erro numérico introduzido, especialmente para problemas não lineares ou para intervalos maiores.<br>